### PR TITLE
[Core] Adding skip test if Application is not added

### PR DESCRIPTION
**Pull request by loumalouomega**
This adds a macro to skip a test in C++ if an app is not imported. improve the run_cpp_test so these tests are run

### CODE REVIEW

File: kratos/python_scripts/run_cpp_tests.py [hunk 1]
```diff
@@ -1,3 +1,19 @@
...
+    import KratosMultiphysics.FluidDynamicsApplication as FluidDynamicsApplication
+except ImportError as e:
+    KM.Logger.PrintWarning("FluidDynamicsApplication is not available")
+try:
+    import KratosMultiphysics.ConvectionDiffusionApplication as ConvectionDiffusionApplication
+except ImportError as e:
+    KM.Logger.PrintWarning("ConvectionDiffusionApplication is not available")
+
+KM.Tester.SetVerbosity(KM.Tester.Verbosity.TESTS_OUTPUTS)
+KM.Tester.RunAllTestCases()
```
Comment (philbucher) [inline kratos/python_scripts/run_cpp_tests.py:3]: yikes
I would leave it as is, this causes more problems than it solves I think

@roigcarlo what do you say?

File: kratos/includes/kernel.h [hunk 2]
```diff
@@ -109,8 +109,17 @@ class KRATOS_API(KRATOS_CORE) Kernel {
...
     void InitializeApplication(KratosApplication& NewApplication) {}
 
-    bool IsImported(std::string ApplicationName) const;
-
+    /**
+     * @brief This method returns if the given application by name is imported
+     * @param rApplicationName The name of the application of interest
+     * @return True if imported, false otherwise
+     */
+    bool IsImported(const std::string& rApplicationName) const;
```
Comment (philbucher) [inline kratos/includes/kernel.h]: I think this can be static

**Review by philbucher — COMMENTED** [review]
cool!
Some comments

File: kratos/includes/kernel.h (hunk 2 above)
Comment (loumalouomega) [reply kratos/includes/kernel.h]: Nope, it cant' I already tried

File: kratos/python_scripts/run_cpp_tests.py (hunk 1 above)
Comment (loumalouomega) [reply kratos/python_scripts/run_cpp_tests.py:3]: In that case the tests depending on other applications will fail

File: kratos/includes/kernel.h (hunk 2 above)
Comment (philbucher) [reply kratos/includes/kernel.h]: well `GetApplicationsList` also needs to be const then
I am not sure why it is not
`application_list` is static

File: kratos/python_scripts/run_cpp_tests.py (hunk 1 above)
Comment (philbucher) [reply kratos/python_scripts/run_cpp_tests.py:3]: they shouldn't fail, they should be skipped

File: kratos/includes/kernel.h (hunk 2 above)
Comment (loumalouomega) [reply kratos/includes/kernel.h]: This was already like this, I only added the documentation

File: kratos/python_scripts/run_cpp_tests.py (hunk 1 above)
Comment (loumalouomega) [reply kratos/python_scripts/run_cpp_tests.py:3]: Yes, technically skip, but what I mean is that in the CI those test will not be run, therefore meaningless

Comment (philbucher) [reply kratos/python_scripts/run_cpp_tests.py:3]: The CI importants every app and then rund all the tests

Comment (philbucher) [reply kratos/python_scripts/run_cpp_tests.py:3]: *imports

Comment (loumalouomega) [reply kratos/python_scripts/run_cpp_tests.py:3]: OK, i trust you

File: kratos/includes/kernel.h (hunk 2 above)
Comment (philbucher) [reply kratos/includes/kernel.h]: I know
But can you please try making them static? I think it would be useful in the future

Comment (loumalouomega) [reply kratos/includes/kernel.h]: I already tried, and gave me compilation errors

Comment (philbucher) [reply kratos/includes/kernel.h]: hmmm
Ok I would like to try myself (and try sth else too)

File: kratos/python/add_kernel_to_python.cpp [hunk 3]
```diff
@@ -123,7 +123,7 @@ void AddKernelToPython(pybind11::module& m)
         .def("InitializeApplication",  [](Kernel& self, KratosApplication& App){ self.Initialize();
         /*RegisterInPythonApplicationVariables(App);*/ }) //&Kernel::InitializeApplication)
         //.def(""A,&Kernel::Initialize)
-        .def("IsImported", &Kernel::IsImported)
+        .def_static("IsImported", &Kernel::IsImported)
```
Comment (loumalouomega) [inline kratos/python/add_kernel_to_python.cpp:126]: Soooo, this was the reason of the fail. I see

**Comment by loumalouomega** [comment]
@philbucher I cannot approve myself...

File: kratos/python/add_kernel_to_python.cpp (hunk 3 above)
Comment (philbucher) [reply kratos/python/add_kernel_to_python.cpp:126]: :)

**Comment by philbucher** [comment]
I know

I would like to let @KratosMultiphysics/technical-committee approve, since I think the dependencies of the tests on C++ level was never discussed before
Esp in relation to #83

**Comment by loumalouomega** [comment]
> I know
> 
> I would like to let @KratosMultiphysics/technical-committee approve, since I think the dependencies of the tests on C++ level was never discussed before
> Esp in relation to #83

Can we merge https://github.com/KratosMultiphysics/Kratos/pull/6681 first then?

**Comment by loumalouomega** [comment]
Ping @philbucher

**Comment by pooyan-dadvand** [comment]
@KratosMultiphysics/technical-committee we have general comments about this PR and the relative tests.

We believe that the cpp tests should be formulated as unit test as possible. Having dependency to other applications is in contrast with this principle and in fact make the test less useful as it gives less information from the location of the error. 

For example instead of using a real element from other application, one should use a dummy element which can be defined in the core/test of Kratos so all using the same thing.

We should emphasize that we want to make CTest compatible version of cpp tests so we can debug and pass it to valgrind without interfacing with python. Having tests with cross application dependency cannot work in this concept. Please consider that we try to avoid having inclusions in cpp between applications.

**Comment by loumalouomega** [comment]
Do I close this then?

**Comment by loumalouomega** [comment]
Do I delete the branch?

**Pull request closed without merging**